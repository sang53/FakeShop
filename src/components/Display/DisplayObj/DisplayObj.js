import Helpers from "../../../Helper/Helpers";

export default class DisplayObj {
  constructor(product, category = false) {
    if (category) {
      this.id = product.category;
      this.title = DisplayObj.capitalise(product.category);
      this.image = product.image;
      this.link = `/categories/${product.category}`;
    } else {
      this.id = product.id;
      this.title = product.title;
      this.image = product.image;
      this.price = Helpers.padDecimal(product.price);
      this.link = `/products/${product.id}`;
    }
  }

  static capitalise(string) {
    return string
      .split(" ")
      .map((word) => word[0].toUpperCase() + word.slice(1))
      .join(" ");
  }

  static getAllCategories(products) {
    const output = [];
    products.forEach((obj) => {
      if (
        !output.some(
          (displayObj) =>
            displayObj.title === DisplayObj.capitalise(obj.category)
        )
      )
        output.push(new DisplayObj(obj, true));
    });
    return output;
  }

  static getCategoryProducts(products, category) {
    return products
      .filter((obj) => obj.category === category)
      .map((obj) => new DisplayObj(obj));
  }
}
