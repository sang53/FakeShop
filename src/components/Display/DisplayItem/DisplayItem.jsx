import PropTypes from "prop-types";
import { Link } from "react-router";
import DisplayObj from "../DisplayObj/DisplayObj";
import { memo } from "react";

const DisplayItem = ({ obj }) => {
  return (
    <Link to={obj.link} key={obj.id} className="product">
      <div className="product-img">
        <img src={obj.image} alt={obj.title} />
      </div>
      <div>
        <div id="title">{obj.title}</div>
        {obj.price && <div>{obj.price}</div>}
      </div>
    </Link>
  );
};

export default memo(DisplayItem);

DisplayItem.propTypes = {
  obj: PropTypes.instanceOf(DisplayObj),
};
