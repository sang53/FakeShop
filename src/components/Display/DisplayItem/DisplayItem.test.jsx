import { test } from "vitest";
import DisplayObj from "../DisplayObj/DisplayObj";
import { mockProducts } from "../../../test/utils/mockProducts";
import { render } from "@testing-library/react";
import DisplayItem from "./DisplayItem";
import { expect } from "vitest";
import { MemoryRouter } from "react-router";

function renderProductItem(category = false) {
  const product = new DisplayObj(mockProducts[0], category);
  return render(
    <MemoryRouter initialEntries={["/"]}>
      <DisplayItem obj={product} />
    </MemoryRouter>
  );
}

test("product item snapshot", () => {
  const { container } = renderProductItem();
  expect(container).toMatchSnapshot();
});

test("category item snapshot", () => {
  const { container } = renderProductItem(true);
  expect(container).toMatchSnapshot();
});
