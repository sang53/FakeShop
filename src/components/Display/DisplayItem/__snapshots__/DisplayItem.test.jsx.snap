// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`category item snapshot 1`] = `
<div>
  <a
    class="product"
    data-discover="true"
    href="/categories/men's clothing"
  >
    <div
      class="product-img"
    >
      <img
        alt="Men's Clothing"
        src="#"
      />
    </div>
    <div>
      <div
        id="title"
      >
        Men's Clothing
      </div>
    </div>
  </a>
</div>
`;

exports[`product item snapshot 1`] = `
<div>
  <a
    class="product"
    data-discover="true"
    href="/products/1"
  >
    <div
      class="product-img"
    >
      <img
        alt="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
        src="#"
      />
    </div>
    <div>
      <div
        id="title"
      >
        Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
      </div>
      <div>
        $109.95
      </div>
    </div>
  </a>
</div>
`;
